Catch errors when calling getObject.

mauritsvanrees opened the pull request:
Can happen when the portal_catalog somehow has a brain for an item that no longer exists:

```
2019-10-03 12:47:13 ERROR Zope.SiteErrorLog
 http://127.0.0.1:8080/intranet/@@revisions-controlpanel
Traceback (innermost last):
  Module ZPublisher.Publish, line 138, in publish
  Module ZPublisher.mapply, line 77, in mapply
  Module ZPublisher.Publish, line 48, in call_object
  Module z3c.form.form, line 233, in __call__
  Module plone.z3cform.fieldsets.extensible, line 59, in update
  Module plone.z3cform.patch, line 30, in GroupForm_update
  Module z3c.form.group, line 145, in update
  Module plone.app.z3cform.csrf, line 21, in execute
  Module z3c.form.action, line 98, in execute
  Module z3c.form.button, line 315, in __call__
  Module z3c.form.button, line 170, in __call__
  Module collective.revisionmanager.browser.views, line 209, in handle_recalculate_stats
  Module collective.revisionmanager.statscache, line 207, in refresh
  Module collective.revisionmanager.statscache, line 109, in _calculate_storage_statistics
  Module collective.revisionmanager.statscache, line 39, in _unrestricted_query_objects
  Module Products.ZCatalog.CatalogBrains, line 100, in getObject
  Module OFS.Traversable, line 300, in unrestrictedTraverse
   - __traceback_info__: (['1', 'knmp-academy'], 'geimporteerd')
KeyError: 'geimporteerd'
```

tschorr commented:
> when the portal_catalog somehow has a brain for an item that no longer exists

Imho this points to catalog inconsistencies, not a bug in `c.revisionmanager`. The correct way to deal with this would be to rebuild the catalog and then recalculate the storage statistics. Because how useful are statistics that are based on an inconsistent catalog? If we absolutely want to deal with this situation in `c.revisionmanager`, I'd prefer showing an error message in the UI to simply logging the error.

mauritsvanrees commented:
(I have rebased and force-pushed the PR branch to fix a conflict in `CHANGES.rst`.)

Correct, this points to a catalog inconsistency. But there is lots of code in Plone that does a `getObject` call and protects this with a `try/except`. Well, apparently not "lots of" now that I search for it, but they are there, mostly in migration code, but also for example in [plone.app.linkintegrity](https://github.com/plone/plone.app.linkintegrity/blob/master/plone/app/linkintegrity/browser/info.py#L63-L65).

Catalog inconsistency just happens. I wrote [collective.catalogcleanup](https://github.com/collective/collective.catalogcleanup) to fix problems in catalogs, and I sometimes run that on sites, but inconsistency can always creep in.

In this case, with my fix I was able to use revisionmanager and I removed some orphans and a few other history versions. And after this cleanup I could again use revisionmanager **without** this fix. In other words, the fix helped me get to a sane state again.

tschorr commented:
> Correct, this points to a catalog inconsistency. But there is lots of code in Plone that does a getObject call and protects this with a try/except. Well, apparently not "lots of" now that I search for it, but they are there, mostly in migration code, but also for example in plone.app.linkintegrity.

I'm not saying that wrapping getObject in a try/except is bad **in general**. It depends on the use case. E.g. being able to display a search results list although the catalog is broken is totally fine. However for `c.revisionmanager` an inconsistent catalog is problematic because it might give you a wrong picture of what is in your histories storage. More specifically, an inconsistent `cmf_uid` index might lead you to think that there is no working copy for an item in the histories list although it actually exists, s. section "Dealing with catalog inconsistencies" in the README.

> In other words, the fix helped me get to a sane state again

But rebuilding the catalog or reindexing `cmf_uid` as suggested in the README would have produced the same result.

In any case I think it is important to notify the user of an inconsistent catalog. Raising an error is the easiest way to do this. If for some reason we are not happy with that, we need to display a message in the UI because we cannot rely on the user to check the logs.

mauritsvanrees commented:
Fair enough. Let’s ignore this PR then. 
I must say I have not read the README that thoroughly...
